File: CS/Reporting_report-storage-for-the-end-user-report-designer-e2704/ZipReportStorage.cs
using System;
using System.IO;
using System.Windows.Forms;
using System.ComponentModel;
using System.IO.Compression;
using System.Collections.Generic;
using DevExpress.Xpo;
using DevExpress.Data.Filtering;        
using DevExpress.XtraReports.UI;                 
using DevExpress.XtraReports.Extensions;           
using DevExpress.Utils.Zip;
// ...

namespace ReportStorageSample {
    class ZipReportStorage : ReportStorageExtension {
        class ZipFilesHelper : IDisposable {
            Stream stream;
            InternalZipFileCollection zipFiles = new InternalZipFileCollection();
            public InternalZipFileCollection ZipFiles {
                get {
                    return zipFiles;
                }
            }
            public ZipFilesHelper(string path) {
                if (File.Exists(path)) {
                    stream = File.OpenRead(path);
                    zipFiles = InternalZipArchive.Open(stream);
                }
            }
            public virtual void Dispose() {
                if (stream != null)
                    stream.Dispose();
            }
        }
        const string fileName = "ReportStorage.zip";
        public ZipReportStorage() {
        }
        string StoragePath {
            get {
                string dirName = Path.GetDirectoryName(Application.ExecutablePath);
                return Path.Combine(dirName, fileName);
            }
        }
        public override bool CanSetData(string url) {
            // Always return true to confirm that the SetData method is available.
            return true;
        }
        public override bool IsValidUrl(string url) {
            return !string.IsNullOrEmpty(url);
        }
        public override byte[] GetData(string url) {
            // Open ZIP archive.
            using (ZipFilesHelper helper = new ZipFilesHelper(StoragePath)) {
                // Read a file with a specified URL from the archive.
                InternalZipFile zipFile = GetZipFile(helper.ZipFiles, url);
                if (zipFile != null)
                    return GetBytes(zipFile);
                return new byte[] { };
            }
        }
        static byte[] GetBytes(InternalZipFile zipFile) {
            return GetBytes(zipFile.FileDataStream, (int)zipFile.UncompressedSize);
        }
        static byte[] GetBytes(Stream stream, int length) {
            byte[] result = new byte[length];
            stream.Read(result, 0, result.Length);
            return result;
        }
        static InternalZipFile GetZipFile(InternalZipFileCollection zipFiles, string url) {
            foreach(InternalZipFile item in zipFiles) {
                if (StringsEgual(item.FileName, url))
                    return item;
            }
            return null;
        }
        static bool StringsEgual(string a, string b) {
            return string.Equals(a, b, StringComparison.OrdinalIgnoreCase);
        }
        public override void SetData(XtraReport report, string url) {
            report.Extensions["StorageID"] = url;
            SaveArchive(url, GetBuffer(report));
        }
        void SaveArchive(string url, byte[] buffer) {
            string tempPath = Path.ChangeExtension(StoragePath, "tmp");
            // Create a new ZIP archive.
            using(InternalZipArchive arch = new InternalZipArchive(tempPath)) {
                // Open a ZIP archive where report files are stored.
                using (ZipFilesHelper helper = new ZipFilesHelper(StoragePath)) {
                    bool added = false;
                    // Copy all report files to a new archive.
                    // Update a file with a specified URL.
                    // If the file does not exist, create it.
                    foreach(InternalZipFile item in helper.ZipFiles) {
                        if (StringsEgual(item.FileName, url)) {
                            arch.Add(item.FileName, DateTime.Now, buffer);
                            added = true;
                        }
                        else
                            arch.Add(item.FileName, DateTime.Now, GetBytes(item));
                    }
                    if (!added)
                        arch.Add(url, DateTime.Now, buffer);
                }
            }
            // Replace the old ZIP archive with the new one.
            if (File.Exists(StoragePath))
                File.Delete(StoragePath);
            File.Move(tempPath, StoragePath);
        }
        byte[] GetBuffer(XtraReport report) {
            using (MemoryStream stream = new MemoryStream()) {
                report.SaveLayout(stream);
                return stream.ToArray();
            }
        }
        public override string GetNewUrl() {
            // Show the report selection dialog and return a URL for a selected report.
            StorageEditorForm form = CreateForm();
            form.textBox1.Enabled = false;
            if (form.ShowDialog() == System.Windows.Forms.DialogResult.OK)
                return form.textBox1.Text;
            return string.Empty;
        }
        StorageEditorForm CreateForm() {
            StorageEditorForm form = new StorageEditorForm();
            foreach (string item in GetUrls())
                form.listBox1.Items.Add(item);
            return form;
        }
        public override string SetNewData(XtraReport report, string defaultUrl) {
            StorageEditorForm form = CreateForm();
            form.textBox1.Text = defaultUrl;
            form.listBox1.Enabled = false;
            // Show the save dialog to get a URL for a new report.
            if (form.ShowDialog() == DialogResult.OK) {
                string url = form.textBox1.Text;
                if (!string.IsNullOrEmpty(url) && !form.listBox1.Items.Contains(url)) {
                    TypeDescriptor.GetProperties(typeof(XtraReport))["DisplayName"].SetValue(report, url);
                    SetData(report, url);
                    return url;
                }
                else {
                    MessageBox.Show("Incorrect report name", "Error",
                        MessageBoxButtons.OKCancel, MessageBoxIcon.Error);
                }
            }
            return string.Empty;
        }
        public override bool GetStandardUrlsSupported(ITypeDescriptorContext context) {
            // Always return true to confirm that the GetStandardUrls method is available.
            return true;
        }
        public override string[] GetStandardUrls(ITypeDescriptorContext context) {
            if (context != null && context.Instance is XRSubreport) {
                XRSubreport xrSubreport = context.Instance as XRSubreport;
                if (xrSubreport.RootReport !=
                    null && xrSubreport.RootReport.Extensions.TryGetValue("StorageID", out storageID)) {
                    List<string> result = GetUrlsCore(CanPassId);
                    return result.ToArray();
                }
            }
            return GetUrls();
        }
        string storageID;
        bool CanPassId(string id) {
            return id != storageID;
        }
        string[] GetUrls() {
            return GetUrlsCore(null).ToArray();
        }
        List<string> GetUrlsCore(Predicate<string> method) {
            List<string> list = new List<string>();
            using (ZipFilesHelper helper = new ZipFilesHelper(StoragePath)) {
                foreach(InternalZipFile item in helper.ZipFiles)
                    if (method == null || method(item.FileName))
                        list.Add(item.FileName);
                return list;
            }
        }
    }
}

File: VB/Reporting_report-storage-for-the-end-user-report-designer-e2704/ZipReportStorage.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.IO
Imports System.Windows.Forms
Imports System.ComponentModel
Imports System.IO.Compression
Imports System.Collections.Generic
Imports DevExpress.Xpo
Imports DevExpress.Data.Filtering
Imports DevExpress.XtraReports.UI
Imports DevExpress.XtraReports.Extensions
Imports DevExpress.Utils.Zip
' ...

Namespace ReportStorageSample
	Friend Class ZipReportStorage
		Inherits ReportStorageExtension
		Private Class ZipFilesHelper
			Implements IDisposable
			Private stream As Stream
			Private zipFiles_Renamed As New InternalZipFileCollection()
			Public ReadOnly Property ZipFiles() As InternalZipFileCollection
				Get
					Return zipFiles_Renamed
				End Get
			End Property
			Public Sub New(ByVal path As String)
				If File.Exists(path) Then
					stream = File.OpenRead(path)
					zipFiles_Renamed = InternalZipArchive.Open(stream)
				End If
			End Sub
			Public Overridable Sub Dispose() Implements IDisposable.Dispose
				If stream IsNot Nothing Then
					stream.Dispose()
				End If
			End Sub
		End Class
		Private Const fileName As String = "ReportStorage.zip"
		Public Sub New()
		End Sub
		Private ReadOnly Property StoragePath() As String
			Get
				Dim dirName As String = Path.GetDirectoryName(Application.ExecutablePath)
				Return Path.Combine(dirName, fileName)
			End Get
		End Property
		Public Overrides Function CanSetData(ByVal url As String) As Boolean
			' Always return true to confirm that the SetData method is available.
			Return True
		End Function
		Public Overrides Function IsValidUrl(ByVal url As String) As Boolean
			Return Not String.IsNullOrEmpty(url)
		End Function
		Public Overrides Function GetData(ByVal url As String) As Byte()
			' Open ZIP archive.
			Using helper As New ZipFilesHelper(StoragePath)
				' Read a file with a specified URL from the archive.
				Dim zipFile As InternalZipFile = GetZipFile(helper.ZipFiles, url)
				If zipFile IsNot Nothing Then
					Return GetBytes(zipFile)
				End If
				Return New Byte() { }
			End Using
		End Function
		Private Shared Function GetBytes(ByVal zipFile As InternalZipFile) As Byte()
			Return GetBytes(zipFile.FileDataStream, CInt(Fix(zipFile.UncompressedSize)))
		End Function
		Private Shared Function GetBytes(ByVal stream As Stream, ByVal length As Integer) As Byte()
			Dim result(length - 1) As Byte
			stream.Read(result, 0, result.Length)
			Return result
		End Function
		Private Shared Function GetZipFile(ByVal zipFiles As InternalZipFileCollection, ByVal url As String) As InternalZipFile
			For Each item As InternalZipFile In zipFiles
				If StringsEgual(item.FileName, url) Then
					Return item
				End If
			Next item
			Return Nothing
		End Function
		Private Shared Function StringsEgual(ByVal a As String, ByVal b As String) As Boolean
			Return String.Equals(a, b, StringComparison.OrdinalIgnoreCase)
		End Function
		Public Overrides Sub SetData(ByVal report As XtraReport, ByVal url As String)
			report.Extensions("StorageID") = url
			SaveArchive(url, GetBuffer(report))
		End Sub
		Private Sub SaveArchive(ByVal url As String, ByVal buffer() As Byte)
			Dim tempPath As String = Path.ChangeExtension(StoragePath, "tmp")
			' Create a new ZIP archive.
			Using arch As New InternalZipArchive(tempPath)
				' Open a ZIP archive where report files are stored.
				Using helper As New ZipFilesHelper(StoragePath)
					Dim added As Boolean = False
					' Copy all report files to a new archive.
					' Update a file with a specified URL.
					' If the file does not exist, create it.
					For Each item As InternalZipFile In helper.ZipFiles
						If StringsEgual(item.FileName, url) Then
							arch.Add(item.FileName, DateTime.Now, buffer)
							added = True
						Else
							arch.Add(item.FileName, DateTime.Now, GetBytes(item))
						End If
					Next item
					If (Not added) Then
						arch.Add(url, DateTime.Now, buffer)
					End If
				End Using
			End Using
			' Replace the old ZIP archive with the new one.
			If File.Exists(StoragePath) Then
				File.Delete(StoragePath)
			End If
			File.Move(tempPath, StoragePath)
		End Sub
		Private Function GetBuffer(ByVal report As XtraReport) As Byte()
			Using stream As New MemoryStream()
				report.SaveLayout(stream)
				Return stream.ToArray()
			End Using
		End Function
		Public Overrides Function GetNewUrl() As String
			' Show the report selection dialog and return a URL for a selected report.
			Dim form As StorageEditorForm = CreateForm()
			form.textBox1.Enabled = False
			If form.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
				Return form.textBox1.Text
			End If
			Return String.Empty
		End Function
		Private Function CreateForm() As StorageEditorForm
			Dim form As New StorageEditorForm()
			For Each item As String In GetUrls()
				form.listBox1.Items.Add(item)
			Next item
			Return form
		End Function
		Public Overrides Function SetNewData(ByVal report As XtraReport, ByVal defaultUrl As String) As String
			Dim form As StorageEditorForm = CreateForm()
			form.textBox1.Text = defaultUrl
			form.listBox1.Enabled = False
			' Show the save dialog to get a URL for a new report.
			If form.ShowDialog() = DialogResult.OK Then
				Dim url As String = form.textBox1.Text
				If (Not String.IsNullOrEmpty(url)) AndAlso (Not form.listBox1.Items.Contains(url)) Then
					TypeDescriptor.GetProperties(GetType(XtraReport))("DisplayName").SetValue(report, url)
					SetData(report, url)
					Return url
				Else
					MessageBox.Show("Incorrect report name", "Error", MessageBoxButtons.OKCancel, MessageBoxIcon.Error)
				End If
			End If
			Return String.Empty
		End Function
		Public Overrides Function GetStandardUrlsSupported(ByVal context As ITypeDescriptorContext) As Boolean
			' Always return true to confirm that the GetStandardUrls method is available.
			Return True
		End Function
		Public Overrides Function GetStandardUrls(ByVal context As ITypeDescriptorContext) As String()
			If context IsNot Nothing AndAlso TypeOf context.Instance Is XRSubreport Then
				Dim xrSubreport As XRSubreport = TryCast(context.Instance, XRSubreport)
				If xrSubreport.RootReport IsNot Nothing AndAlso xrSubreport.RootReport.Extensions.TryGetValue("StorageID", storageID) Then
					Dim result As List(Of String) = GetUrlsCore(AddressOf CanPassId)
					Return result.ToArray()
				End If
			End If
			Return GetUrls()
		End Function
		Private storageID As String
		Private Function CanPassId(ByVal id As String) As Boolean
			Return id <> storageID
		End Function
		Private Function GetUrls() As String()
			Return GetUrlsCore(Nothing).ToArray()
		End Function
		Private Function GetUrlsCore(ByVal method As Predicate(Of String)) As List(Of String)
			Dim list As New List(Of String)()
			Using helper As New ZipFilesHelper(StoragePath)
				For Each item As InternalZipFile In helper.ZipFiles
					If method Is Nothing OrElse method(item.FileName) Then
						list.Add(item.FileName)
					End If
				Next item
				Return list
			End Using
		End Function
	End Class
End Namespace
